ideafast/middleware-services
# Dockerise data transfer pipeline

## What Changed?

1. Added dockerfile for datatransfer middleware
2. Modified docker-compose file to run all services for pipeline
3. Refactored `main.py` by pulling out `dags` to their own folder. This is helpful when running batch jobs
4. Refactored dreem DAG (`drm.py`) to operate on a per-patient level, which reduces the quantity of data stored on the server at once.
5. Running dreem pipeline as batchjob that invokes a DAG via the docker file. This is a temporary solution as we do not know which devices will make it into COS, so no need to fancy airflow (or celery) solution yet. 
6. Added solution to use development environmental variables when they are provided, e.g., for local development.
7. Refactored how `-meta.json` is stored to share logging/logic between DAGs.
8. Added error handling and logging to dreem library. 

Notes:

- `Dtransfer.Dockerfile` produces a large image (346MB) I can optimise this to ~100MB if necessary, but results in longer build times.
- We should pull out error handling a level higher, but imho should do this in another PR.
- I've not bothered updating `cli.py` yet to invoke docker build as we should pull out datatransfer to a new repo before then.
- Currently runs every minute: we probably should modify `Dtransfer.Dockerfile` so that setting when the pipeline is run is configurable, e.g., nightly.
 
## How to Test?

0. Modify code: Add `all_records = all_records[-9::3]` after the `dreem_api` call so that the pipeline operates on 3 files.
1. Build images: from the root folder `docker build -t "middleware" . ` to build the middleware image (consumer) and `docker build -f "Dtransfer.Dockerfile" -t "dtransfer" .` to build datatransfer image.
2. Run Services: run `docker-compose up -d`
3. Observe logs: run `docker logs -f dtransfer` to see the logs and note that it runs every minute (current config).
4. Check DMP: check hackathon dataset on DMP to see data uploaded

Closes: #16 and #29 (partially)
